File: node.js/백준/Gold/7576. 토마토/토마토.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf8').toString().trim().split('\n');
const [N, M] = input[0].trim().split(' ').map(Number);
const map = Array.from({length: M}, () => Array(N).fill(0));
const visited = Array.from({length: M}, () => Array(N).fill(false));

// 시작 지점 좌표
let start_position = [];
// 안 익은 토마토 개수
let nomato = 0;
for (let i = 1; i <= M; i++) {
    const row = input[i].trim().split(' ').map(Number);
    for (let j = 0; j < N; j++) {
        map[i - 1][j] = row[j];
        if (row[j] === 1) {
            // 시작 지점 좌표를 배열에 저장
            start_position.push([i - 1, j]);
        } else if (row[j] === 0) {
            // 안 익은 토마토 카운팅
            nomato++;
        }
    }
}

// 모든 토마토가 익어있는 상태 예외 처리
if (nomato === 0) {
    console.log(0);
} else {
    // 최대 depth = 최소 날짜
    const day = BFS(start_position);

    // 안 익은 토마토가 1개라도 있으면 -1, 그렇지 않으면 day 출력
    const result = nomato ? -1 : day;
    console.log(result);
}

function BFS(start_position) {
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];
    let maxDepth = 0;
    const queue = [];
    let index = -1;
    for (const [x, y] of start_position) {
        // 시작 지점에서 한 번에 BFS 실행
        visited[x][y] = true;
        queue.push([x, y, 0]);
    }
    while (queue.length > 0) {
        index++;
        const [currentX, currentY, depth] = queue[index];
        // JavaScript Array의 shift() 메서드는 O(n)이라 느림!
        // 따라서 인덱스로 관리
        // const [currentX, currentY, depth] = queue.shift();
        maxDepth = Math.max(maxDepth, depth);
        for (let i = 0; i < 4; i++) {
            const nextX = dx[i] + currentX;
            const nextY = dy[i] + currentY;
            if (isInside(nextX, nextY) && !visited[nextX][nextY] && map[nextX][nextY] === 0) {
                visited[nextX][nextY] = true;
                map[nextX][nextY] = 1;
                nomato--;
                queue.push([nextX, nextY, depth + 1]);
            }
        }
        if (index >= queue.length - 1) {
            break;
        }
    }
    return maxDepth;
}

function isInside(x, y) {
    return x >= 0 && x <= M - 1 && y >= 0 && y <= N - 1;
}

File: Java/백준/Gold/7569. 토마토/토마토.java
import java.io.*;
import java.util.*;

public class Main {
    static int M;
    static int N;
    static int H;
    static int[][][] map;
    static Queue<int[]> queue;
    static int nomato;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());
        map = new int[H][N][M];
        queue = new LinkedList<>();
        nomato = 0;
        for (int i = 0; i < H; i++) {
            for (int j = 0; j < N; j++) {
                st = new StringTokenizer(br.readLine());
                for (int k = 0; k < M; k++) {
                    map[i][j][k] = Integer.parseInt(st.nextToken());
                    if (map[i][j][k] == 1) {
                        queue.add(new int[]{i, j, k, 0});
                    } else if (map[i][j][k] == 0) {
                        nomato++;
                    }
                }
            }
        }
        if (nomato == 0) {
            System.out.println(0);
        } else {
            int day = BFS();
            if (nomato == 0) {
                System.out.println(day);
            } else {
                System.out.println(-1);
            }
        }
    }

    public static int BFS() {
        int[] dx = {-1, 1, 0, 0, 0, 0};
        int[] dy = {0, 0, -1, 1, 0, 0};
        int[] dz = {0, 0, 0, 0, -1, 1};
        int maxDepth = 0;
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int currentX = current[0];
            int currentY = current[1];
            int currentZ = current[2];
            int depth = current[3];
            maxDepth = Math.max(maxDepth, depth);
            for (int i = 0; i < 6; i++) {
                int nextX = dx[i] + currentX;
                int nextY = dy[i] + currentY;
                int nextZ = dz[i] + currentZ;
                if (isInside(nextX, nextY, nextZ) && map[nextX][nextY][nextZ] == 0) {
                    map[nextX][nextY][nextZ] = 1;
                    queue.offer(new int[]{nextX, nextY, nextZ, depth + 1});
                    nomato--;
                }
            }
        }
        return maxDepth;
    }

    public static boolean isInside(int x, int y, int z) {
        return x >= 0 && x <= H - 1 && y >= 0 && y <= N - 1 && z >= 0 && z <= M - 1;
    }
}